storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3108,10 +3108,1655 @@ buf_block_t *fsp_get_latched_xdes_page(
+  dberr_t defragment_system_tables() noexcept
+  {
+    dberr_t err= defragment_table(dict_sys.sys_tables);
+    if (err == DB_SUCCESS)
+      err= defragment_table(dict_sys.sys_columns);
+    if (err == DB_SUCCESS)
+      err= defragment_table(dict_sys.sys_indexes);
+    if (err == DB_SUCCESS)
+      err= defragment_table(dict_sys.sys_fields);
+    if (err == DB_SUCCESS)
+      err= defragment_table(dict_sys.sys_foreign);
+    if (err == DB_SUCCESS)
+      err= defragment_table(dict_sys.sys_foreign_cols);
+    if (err == DB_SUCCESS)
+      err= defragment_table(dict_sys.sys_virtual);
+    return err;
+  }
+
+  /** @return extent which replaces the later extent
+  or same extent if there is no replacement exist */
+  uint32_t get_new_extent(uint32_t old_extent) const noexcept
+  {
+    auto it= m_extent_map.find(old_extent);
+    if (it != m_extent_map.end())
+      return it->second;
+    return old_extent;
+  }
+
+  /** @return state for the given extent */
+  uint32_t get_state(uint32_t extent) noexcept
+  {
+    return m_extent_info[extent];
+  }
+};
+
+#ifdef UNIV_DEBUG
+/** Validate the system tablespace list */
+__attribute__((warn_unused_result))
+dberr_t fsp_tablespace_validate(fil_space_t *space)
+{
+  /* Validate all FSP list in system tablespace */
+  mtr_t local_mtr;
+  dberr_t err= DB_SUCCESS;
+  local_mtr.start();
+  if (buf_block_t *header= fsp_get_header(
+        space, &local_mtr, &err))
+  {
+    flst_validate(header, FSP_FREE + FSP_HEADER_OFFSET, &local_mtr);
+    flst_validate(header, FSP_FREE_FRAG + FSP_HEADER_OFFSET,
+                  &local_mtr);
+    flst_validate(header, FSP_HEADER_OFFSET + FSP_FULL_FRAG,
+                  &local_mtr);
+    flst_validate(header, FSP_HEADER_OFFSET + FSP_SEG_INODES_FULL,
+                  &local_mtr);
+    flst_validate(header, FSP_HEADER_OFFSET + FSP_SEG_INODES_FREE,
+                  &local_mtr);
+  }
+  local_mtr.commit();
+  return err;
+}
+
+/** Validate the system tablespace list */
+__attribute__((warn_unused_result))
+dberr_t fseg_validate(fil_space_t *space, dict_index_t *index)
+{
+  /* Validate all FSP list in system tablespace */
+  mtr_t local_mtr;
+  dberr_t err= DB_SUCCESS;
+  local_mtr.start();
+
+  buf_block_t *root= btr_root_block_get(index, RW_SX_LATCH,
+                                        &local_mtr, &err);
+  if (UNIV_UNLIKELY(!root))
+  {
+err_exit:
+    local_mtr.commit();
+    return err;
+  }
+
+  fseg_header_t *seg_header=
+    root->page.frame + PAGE_HEADER + PAGE_BTR_SEG_TOP;
+
+  buf_block_t *iblock;
+  fseg_inode_t *inode= fseg_inode_try_get(seg_header, 0, 0, &local_mtr,
+                                          &iblock, &err);
+  if (!inode)
+    goto err_exit;
+
+  uint16_t i_offset= uint16_t(inode - iblock->page.frame);
+
+  flst_validate(iblock, uint16_t(i_offset + FSEG_FREE), &local_mtr);
+  flst_validate(iblock, uint16_t(i_offset + FSEG_NOT_FULL), &local_mtr);
+  flst_validate(iblock, uint16_t(i_offset + FSEG_FULL), &local_mtr);
```

Comment:
Didn’t we want to execute this validation in all builds (not only `CMAKE_BUILD_TYPE=Debug`), and not crash on an error? The idea would be to validate everything before starting to modify the system tablespace.